UnityProject/Assets/Scripts/Systems/Explosions/ExplosionComponent.cs

```diff
@@ -114,17 +142,87 @@ private IEnumerator ExplosionRoutine(Matrix matrix)
 			Destroy(gameObject);
 		}
 
-		public IEnumerator TimedFireEffect(Vector3Int position, float time, TileChangeManager tileChangeManager)
+		public IEnumerator TimedEffect(Vector3Int position, float time, bool isEmp, TileChangeManager tileChangeManager)
 		{
-			//Dont do fire if already fire
-			if(tileChangeManager.MetaTileMap.HasOverlay(position, TileType.Effects, "Fire")) yield break;
+			string effectName;
+			OverlayType effectOverlayType;
+			if (isEmp)
+			{
+				effectName = "EMPEffect";
+				effectOverlayType = OverlayType.EMP;
+			}
+			else
+			{
+				effectName = "Fire";
+				effectOverlayType = OverlayType.Fire;
+			}
+			//Dont add effect if it is already there
+			if(tileChangeManager.MetaTileMap.HasOverlay(position, TileType.Effects, effectName)) yield break;
 
-			tileChangeManager.MetaTileMap.AddOverlay(position, TileType.Effects, "Fire");
+			tileChangeManager.MetaTileMap.AddOverlay(position, TileType.Effects, effectName);
 			yield return WaitFor.Seconds(time);
-			tileChangeManager.MetaTileMap.RemoveOverlaysOfType(position, LayerType.Effects, OverlayType.Fire);
+			tileChangeManager.MetaTileMap.RemoveOverlaysOfType(position, LayerType.Effects, effectOverlayType);
+		}
+
+		private void EmpThing(GameObject thing, int EmpStrength)
+		{
+			if(thing != null)
+			{
+				if (isEmpAble(thing))
```

Comment:
you did an interface that identifies everything that is EMPable, you should check for it instead.

pseudo code

```cs
foreach (var thing in explosionRangeObjects)
{
    if (thing.TryGetComponent<IEmpable>(out var empable)
    {
        empable.OnEmp(empStrenght)
    }
}
```

This way you avoid getting so many different components, I think explosions do affect stuff inside inventories, correct me if I am wrong.

Reply:
Regular explosion don't affect things in inventory. Also, I think I need to know to which gameObject IEmpAble is attached for storage checks. Can I just make method with default implementation in IEmpAble, something like GetGameObject() for it?

Reply:
Can I resolve it?

Reply:
what's the difference between ``ExplosionComponent`` and ``ExplosionNode`` ? They seem to have duplicated code

Reply:
ExplosionComponent is made for spawning explosion prefabs, it has different algorithm for making explosion shape and it also applies visual overlay. Explosion is same but can be triggered StartExplosion. I think we need to make single script for explosions but, again, I think Ill do this in separate PR

Reply:
aight, fair enough